src/AdoNet/Orleans.Streaming.AdoNet/AdoNetQueueAdapterFactory.cs

```diff
@@ -0,0 +1,122 @@
+using System.Threading;
+using Microsoft.Extensions.Hosting;
+
+namespace Orleans.Streaming.AdoNet;
+
+internal class AdoNetQueueAdapterFactory : IQueueAdapterFactory
+{
+    public AdoNetQueueAdapterFactory(string name, AdoNetStreamOptions streamOptions, ClusterOptions clusterOptions, SimpleQueueCacheOptions cacheOptions, HashRingStreamQueueMapperOptions hashOptions, StreamPullingAgentOptions agentOptions, ILoggerFactory loggerFactory, IHostApplicationLifetime lifetime, IServiceProvider serviceProvider)
+    {
+        _name = name;
+        _streamOptions = streamOptions;
+        _clusterOptions = clusterOptions;
+        _cacheOptions = cacheOptions;
+        _agentOptions = agentOptions;
+        _lifetime = lifetime;
+        _serviceProvider = serviceProvider;
+
+        _streamQueueMapper = new HashRingBasedStreamQueueMapper(hashOptions, name);
+        _cache = new SimpleQueueAdapterCache(cacheOptions, name, loggerFactory);
+        _adoNetQueueMapper = new AdoNetStreamQueueMapper(_streamQueueMapper);
+    }
+
+    private readonly string _name;
+    private readonly AdoNetStreamOptions _streamOptions;
+    private readonly ClusterOptions _clusterOptions;
+    private readonly SimpleQueueCacheOptions _cacheOptions;
+    private readonly StreamPullingAgentOptions _agentOptions;
+    private readonly IHostApplicationLifetime _lifetime;
+    private readonly IServiceProvider _serviceProvider;
+
+    private readonly HashRingBasedStreamQueueMapper _streamQueueMapper;
+    private readonly SimpleQueueAdapterCache _cache;
+    private readonly AdoNetStreamQueueMapper _adoNetQueueMapper;
+
+    private RelationalOrleansQueries _queries;
+
+    /// <summary>
+    /// Unfortunate implementation detail to account for lack of async lifetime.
+    /// Ideally this concern will be moved upstream so this won't be needed.
+    /// </summary>
+    private readonly SemaphoreSlim _semaphore = new(1);
+
+    /// <summary>
+    /// Ensures queries are loaded only once while allowing for recovery if the load fails.
+    /// </summary>
+    private ValueTask<RelationalOrleansQueries> GetQueriesAsync()
+    {
+        // attempt fast path
+        return _queries is not null ? new(_queries) : new(CoreAsync());
+
+        // slow path
+        async Task<RelationalOrleansQueries> CoreAsync()
+        {
+            await _semaphore.WaitAsync(_lifetime.ApplicationStopping);
```

Comment:
Do you want to pass in a timeout here too?

Reply:
Thanks, safety timeout added.